src/ImageSharp/Formats/Png/PngDecoderCore.cs

```diff
@@ -1206,23 +1206,30 @@ private bool TryReadChunk(out PngChunk chunk)
 
             PngChunkType type = this.ReadChunkType();
 
-            // NOTE: Reading the Data chunk is the responsible of the caller
             // If we're reading color metadata only we're only interested in the IHDR and tRNS chunks.
             // We can skip all other chunk data in the stream for better performance.
-            if (type == PngChunkType.Data || (this.colorMetadataOnly && type != PngChunkType.Header && type != PngChunkType.Transparency))
+            if (this.colorMetadataOnly && type != PngChunkType.Header && type != PngChunkType.Transparency)
             {
                 chunk = new PngChunk(length, type);
 
                 return true;
             }
 
+            long pos = this.currentStream.Position;
             chunk = new PngChunk(
                 length: length,
                 type: type,
                 data: this.ReadChunkData(length));
 
             this.ValidateChunk(chunk);
 
+            // Restore the stream position for IDAT chunks, because it will be decoded later and
+            // was only read to verifying the CRC is correct.
+            if (type == PngChunkType.Data)
+            {
+                this.currentStream.Position = pos;
+            }
```

Comment:
Why do we need to read the IDAT chunk twice? Upon rewinding the stream, this means creating & filling `PngChunk.Data` twice. Can't we avoid this?

Reply:
I dont see how we can avoid reading the stream twice. We need to read the stream completely once to be able to calculate the CRC. Then we are sure the byte stream is not corrupted and can start decoding it in `ZlibInflateStream`. Otherwise we would need to introduce PNG specific logic into `ZlibInflateStream`, which i dont think would be good. 
Also the compressed stream can span over multiple IDAT chunks, which would make it even more complicated.

Note also the `PngChunk.Data` is not filled twice, just once. Before this PR the `PngChunk.Data` was left empty.
So we have now with this PR an additional allocation of the size of `IDAT` chunk.

Reply:
Does libpng check it? I know it’s optional but haven’t had the chance to dig through the code

Reply:
yes, libpng does check it

Reply:
One way to possible avoid a large allocation here for the `IDAT` chunk just to calculate the CRC could be to read the `IDAT` data in smaller portions from the stream (lets say 8192 bytes) and calculate the CRC on those smaller chunk's of data.

Is it worth introducing such special handling for calculating the CRC of `IDAT`?

Reply:
ZlibInflateStream operates over `BufferedReadStream`, so we can consider making CRC32 optional part of `BufferedReadStream` buffer fills, although it's definitely not a 2.0 optimization, I would open a separate issue to investigate this.

> read the `IDAT` data in smaller portions from the stream (lets say 8192 bytes) and calculate the CRC on those smaller chunk's of data.

If my understanding is correct, PNG-s with single very large IDAT chunks are rare, so it's probably not worth it, I would better investigate the `BufferedReadStream` approach.